blockstore/splitstore/splitstore.go

```diff
@@ -458,266 +616,528 @@ func (s *SplitStore) HeadChange(_, apply []*types.TipSet) error {
 	return nil
 }
 
-func (s *SplitStore) warmup(curTs *types.TipSet) {
-	epoch := curTs.Height()
-
-	batchHot := make([]blocks.Block, 0, batchSize)
-	batchSnoop := make([]cid.Cid, 0, batchSize)
-
-	count := int64(0)
-	err := s.chain.WalkSnapshot(context.Background(), curTs, 1, s.skipOldMsgs, s.skipMsgReceipts,
-		func(cid cid.Cid) error {
-			count++
+// transactionally protect incoming tipsets
+func (s *SplitStore) protectTipSets(apply []*types.TipSet) {
+	s.txnLk.RLock()
+	defer s.txnLk.RUnlock()
 
-			has, err := s.hot.Has(cid)
-			if err != nil {
-				return err
-			}
+	if !s.txnActive {
+		return
+	}
 
-			if has {
-				return nil
-			}
+	var cids []cid.Cid
+	for _, ts := range apply {
+		cids = append(cids, ts.Cids()...)
+	}
 
-			blk, err := s.cold.Get(cid)
-			if err != nil {
-				return err
-			}
+	s.trackTxnRefMany(cids)
+}
 
-			batchHot = append(batchHot, blk)
-			batchSnoop = append(batchSnoop, cid)
+// transactionally protect a view
+func (s *SplitStore) protectView(c cid.Cid) {
+	s.txnLk.RLock()
+	defer s.txnLk.RUnlock()
 
-			if len(batchHot) == batchSize {
-				err = s.tracker.PutBatch(batchSnoop, epoch)
-				if err != nil {
-					return err
-				}
-				batchSnoop = batchSnoop[:0]
+	if s.txnActive {
+		s.trackTxnRef(c)
+	}
 
-				err = s.hot.PutMany(batchHot)
-				if err != nil {
-					return err
-				}
-				batchHot = batchHot[:0]
-			}
+	s.txnViewsMx.Lock()
+	s.txnViews++
+	s.txnViewsMx.Unlock()
+}
 
-			return nil
-		})
+func (s *SplitStore) viewDone() {
+	s.txnViewsMx.Lock()
+	defer s.txnViewsMx.Unlock()
 
-	if err != nil {
-		log.Errorf("error warming up splitstore: %s", err)
-		return
+	s.txnViews--
+	if s.txnViews == 0 && s.txnViewsWaiting {
+		s.txnViewsCond.Signal()
```

Comment:
There's no reason to use `Signal` instead of `Broadcast` here. It _shouldn't_ matter, but think of `Signal` as an optimization when you _only_ want to wake up one waiter (when you expect multiple).

Reply:
Not critical... but still.

Reply:
ok, I'll fix it in the open gc pr to use Broadcast.

Reply:
fixed in https://github.com/filecoin-project/lotus/pull/6728/commits/86d7d1472d37003a63d08bb451dc70bb5a3d5f1d